### PR TITLE
Fixes #24259 - Add canned admin role

### PR DESCRIPTION
**Pull request by xprazak2**
The new role has create, view, destroy permissions for organizations and view, edit for settings.

### CODE REVIEW

**Comment by ares** [comment]
it seems Katello tests failures are related, could you please take a look?

**Comment by xprazak2** [comment]
[test katello]

**Comment by lzap** [comment]
This is quite a change, can you start a thread on the list about what's this is all about? We also need to write docs after this is merged.

**Comment by ares** [comment]
It seems pretty straightforward to me, but I do agree the documentation for new role would be helpful and of course a demo once it's merged :-) AFAICT all it does is a permission for Settings page (today only admin can access it) and seeds a new role with these permissions.

EDIT: please don't get me wrong, sharing the info about the work on discourse can do no harm :-)

**Comment by lzap** [comment]
You are right the ticket describes this clearly "It was requested that we ship a new default role - Canned admin. This role is supposed to create new organizations without access to entities within those organizations and manage system wide settings." thanks.

File: app/models/role.rb
Comment (ares) [inline app/models/role.rb]: Could we use scoped instead of where('0 = 0')?

File: app/models/role.rb
Comment (ares) [inline app/models/role.rb]: Could you pls add parentheses so we don't have to think abour operators priority?

File: app/models/role.rb
Comment (ares) [inline app/models/role.rb]: based on the name, I'm not entirely sure what it's supposed to do, also shouldn't it respect roles assigned via usergroups?

File: app/models/user.rb
Comment (ares) [inline app/models/user.rb]: I'm not sure we should try to limit this user too much. Whoever has create permission or create roles permission can basically escalate his own permissions. The same applies for assign_orgs, assign_locs. Also this user can disable login entirely via settings. Preventing escalation is probably a good idea but maybe we should add it for all users and as a separate PR?

**Comment by mmoll** [comment]
@xprazak2 test failures related

**Review by ares — CHANGES_REQUESTED** [review]
few small comments, one concern about handling escalation, then I'm happy to do testing

File: app/models/role.rb
Comment (xprazak2) [reply app/models/role.rb]: I changed the name, it now shows the roles from usergroups.

File: app/models/user.rb
Comment (xprazak2) [reply app/models/user.rb]: Disabling login is a way to get to the resources you otherwise would not be able to. But the fact that someone can disable the whole user authentication system should not be a reason why our roles and permissions do not work properly when enabled.

I think preventing an escalation is in this case a necessity. One of the requirements was that the role will be able to assign roles without having them and without self assignment prevented, user having it could modify itself to access almost anything.

You are right that there are cases that need to be fixed but those are existing bugs and I would prefer not to knowingly add more of them. When I tested, I was not able to escalate my roles/permissions as 'Canned admin', so I kept the escalation check for now but I can remove it if you think it should be added together with the fixes for the remaining cases.

**Comment by san7ket** [comment]
One Failed Scenario-
As a Canned Admin I am able to create new user with the Roles I don't have, but after Creation of those new users, I am unable to edit them, or delete them.
In the Screenshot, you see no actions can be performed on the user by canned admin, created by canned admin.
cannedadmin

File: app/models/user.rb
Comment (ares) [reply app/models/user.rb]: ok, sounds good and we can keep it in, I'll focus on this while testing too, but I think if system admin can create users and change other users permissions, she can escalate through second account easily

File: db/seeds.d/020-permissions_list.rb
Comment (ares) [inline db/seeds.d/020-permissions_list.rb]: Should this be generic permission? Otherwise users could do filtering e.g. on role name, which this patch does not seem to address. We only check whether user has this permission. In other words, resource type shoulf be nil?

File: app/models/user.rb
Comment (ares) [inline app/models/user.rb]: I think if user can escalate roles, we shouldn't forbid him to escalate his roles. He can always use another account for escalation.

File: db/seeds.d/020-roles_list.rb
Comment (ares) [inline db/seeds.d/020-roles_list.rb]: also usergroups and auth sources, roles are required

**Review by ares — CHANGES_REQUESTED** [review]
few more things after reviewing together

**Comment by xprazak2** [comment]
@ares, updated. I removed the escalation check for self and added permissions for the remaining resources in Administer tab.

File: db/seeds.d/020-roles_list.rb
Comment (ares) [inline db/seeds.d/020-roles_list.rb]: Does not block testing but the description should mention other added resources and potentially a fact, this is a powerful role that can potentially get access to all resources.

**Review by ares — CHANGES_REQUESTED** [review]

File: app/models/user.rb
Comment (ares) [inline app/models/user.rb]: This is now same as `can_escalate_for_usergroup?`, if you want to keep separate methods for future, I suggest you add third method `can_escalate?` and call it from both to deduplicate.

Alternative, and in fact preferred, is just keep single method and also single `assignable_to` scope. In fact this was the idea for `for_current_user` scope which you extracted to `delegate_to`. This is no longer needed and we can keep `for_current_user`, just change the where("0=0") to `givable` as you did in your method.

File: db/seeds.d/020-roles_list.rb
Comment (ares) [inline db/seeds.d/020-roles_list.rb]: sorry to mention quite late, but "canned admin" is not great name, all roles are canned, how about just "Admin" or "System admin"?

File: app/models/role.rb
Comment (ares) [inline app/models/role.rb]: please change the name of the role, see below for suggestions, the constant may remain the same but preferably it should also be changed to keep the code readable

**Review by ares — CHANGES_REQUESTED** [review]
Leaving few more code related comments, nothing big I hope.

One thing from testing - canned admin user can't create filter with some permissions, e.g. acess dashboard or view_archtiectures, it seems view_domains works, most likely related to taxonomies?

Otherwise I think this works fine! Once it gets in, we should update our manual with the role description.

**Comment by ares** [comment]
test failures seems related

**Comment by xprazak2** [comment]
I changed canned admin to system admin, the role should now be able to create filters for resources that are not taxable.

**Comment by ares** [comment]
tests failures are related and need to be fixed, code looks good now, I'll retest but looks promising

**Comment by san7ket** [comment]
When a System role tries to edit any filter on any role, it may have created or not. We get a 404 
Filter not found Please try to update your request

```
07:27:51 rails.1   | 2018-09-25T07:27:51 [I|app|] Started GET "/filters/306/edit?role_id=24-System+Admin+Super+admin" for 127.0.0.1 at 2018-09-25 07:27:51 +0000
07:27:51 rails.1   | 2018-09-25T07:27:51 [I|app|020cd] Processing by FiltersController#edit as HTML
07:27:51 rails.1   | 2018-09-25T07:27:51 [I|app|020cd]   Parameters: {"role_id"=>"24-System Admin Super admin", "id"=>"306"}
07:27:52 rails.1   | 2018-09-25T07:27:52 [I|app|020cd] Current user set to foreman (regular)
07:27:52 rails.1   | 2018-09-25T07:27:52 [D|app|020cd] Current location set to Default Location
07:27:52 rails.1   | 2018-09-25T07:27:52 [D|app|020cd] Current organization set to Default Organization
07:27:52 rails.1   | 2018-09-25T07:27:52 [D|app|020cd] not found: Couldn't find Filter with 'id'=306 [WHERE "filters"."role_id" = $1 AND (1=0)]
07:27:52 rails.1   | 2018-09-25T07:27:52 [I|app|020cd]   Rendering common/404.html.erb within layouts/application
07:27:52 rails.1   | 2018-09-25T07:27:52 [I|app|020cd]   Rendered common/404.html.erb within layouts/application (2.3ms)
07:27:52 rails.1   | 2018-09-25T07:27:52 [I|app|020cd]   Rendered layouts/_application_content.html.erb (3.0ms)
07:27:52 rails.1   | 2018-09-25T07:27:52 [I|app|020cd]   Rendering layouts/base.html.erb
07:27:52 rails.1   | 2018-09-25T07:27:52 [I|app|020cd]   Rendered layouts/base.html.erb (536.8ms)
07:27:52 rails.1   | 2018-09-25T07:27:52 [I|app|020cd] Completed 404 Not Found in 649ms (Views: 559.2ms | ActiveRecord: 17.8ms)
07:27:52 rails.1   |  | 
07:27:52 rails.1   |  | 
07:27:52 rails.1   | 2018-09-25T07:27:52 [I|app|] Started GET "/filters/306/edit?role_id=24-System+Admin+Super+admin" for 127.0.0.1 at 2018-09-25 07:27:52 +0000
07:27:52 rails.1   | 2018-09-25T07:27:52 [I|app|feb8c] Processing by FiltersController#edit as HTML
07:27:52 rails.1   | 2018-09-25T07:27:52 [I|app|feb8c]   Parameters: {"role_id"=>"24-System Admin Super admin", "id"=>"306"}
07:27:52 rails.1   | 2018-09-25T07:27:52 [I|app|feb8c] Current user set to foreman (regular)
07:27:52 rails.1   | 2018-09-25T07:27:52 [D|app|feb8c] Current location set to Default Location
07:27:52 rails.1   | 2018-09-25T07:27:52 [D|app|feb8c] Current organization set to Default Organization
07:27:52 rails.1   | 2018-09-25T07:27:52 [D|app|feb8c] not found: Couldn't find Filter with 'id'=306 [WHERE "filters"."role_id" = $1 AND (1=0)]
07:27:52 rails.1   | 2018-09-25T07:27:52 [I|app|feb8c]   Rendering common/404.html.erb within layouts/application
07:27:52 rails.1   | 2018-09-25T07:27:52 [I|app|feb8c]   Rendered common/404.html.erb within layouts/application (9.1ms)
07:27:52 rails.1   | 2018-09-25T07:27:52 [I|app|feb8c]   Rendered layouts/_application_content.html.erb (1.6ms)
07:27:52 rails.1   | 2018-09-25T07:27:52 [I|app|feb8c]   Rendering layouts/base.html.erb
```

**Comment by xprazak2** [comment]
Fixing rubocop issues and editing a filter.

**Comment by xprazak2** [comment]
I added as separate commit which prevents users from modifying filters if they do not have `:edit_roles` permission, but the whole solution is very hacky. I think it would be better to do this separately and turn filters into a proper nested resource.

**Comment by johnsonm325** [comment]
@xprazak2 Katello has a bug that is blocked by this fix (https://projects.theforeman.org/issues/25105). Once the admin role is added, we should be able to tell the user what permissions they are missing in the 403 response.

Just wanted to make you aware of what we were doing. Thanks for your work on this.

**Comment by ares** [comment]
@johnsonm325 are you sure this will help in your case? I think Foreman already makes this available, but bastion does not read/display such information on 403 page.

**Review by ares — APPROVED** [review]

**Review by ares — APPROVED** [review]
Thanks @xprazak2, works great now. Mind to add a note to our [manual](https://www.theforeman.org/manuals/1.18/index.html#4.1.2RolesandPermissions) about this new role? Also `escalate_roles` should probably be mentioned there.

Test failures unrelated

Merging, definitely something worth demoing :-)

**Comment by mmoll** [comment]
It seems that failing test on MySQL and Ruby 2.5 is consistent (seen in the last test_develop runs since this merge).

**Comment by ares** [comment]
Thanks for tracing that Michael! @xprazak2 could you please take a look? It seems mysql adapter renders SQL with " instead of ` or something...

```
Expected "SELECT `roles`.* FROM `roles` WHERE `roles`.`builtin` = 0 ORDER BY `roles`.`name` ASC" to include "WHERE \"roles\".\"builtin\" = 0". (Minitest::Assertion)
/var/lib/workspace/workspace/test_develop_pr_core/database/mysql/ruby/2.5/slave/fast/test/models/role_test.rb:212    
```

**Comment by johnsonm325** [comment]
> @johnsonm325 are you sure this will help in your case? I think Foreman already makes this available, but bastion does not read/display such information on 403 page.

@ares Yes, I just updated my foreman repo and it gave me the 403 for view_settings but actually provided the missing permission (view_settings).

**Comment by ares** [comment]
ok then, glad it helped :-)

**Pull request merged**
